### PR TITLE
[Hetero][RFC] Heterogeneous graph Python interfaces & Message Passing

### PR DESCRIPTION
**Pull request by BarclayII**
## Description
This PR implements

* the Python interfaces to heterogeneous graphs
* message passing by edge type.

Note that although this is still a draft, reviewers could start going through it as it is already pretty big.

### API changes to `HeteroGraph`

* All graph query API arguments are prepended with a type argument.  For instance,
  * `g.has_nodes(v) -> g.has_nodes(node_type, v)`
  * `g.in_edges(v) -> g.in_edges(edge_type, v)`
* `apply_nodes`, `apply_edges`, and `group_apply_edges` takes in a dictionary of node/edge types and functions, as well as an optional dictionary of node/edge types and node/edge ID lists.
* Message passing routines like `send_and_recv` and `update_all` currently only supports the slice syntax:
  ```python
  g[srctype, etype, dsttype].update_all(msg_func, reduce_func, apply_node_func)
  g[srctype, etype, dsttype].send_and_recv(edges, msg_func, reduce_func, apply_node_func)
  ```
  We defer the support of dictionary of functions to later PRs due to the necessity of cross-type reducers.

### Other changes

* `send_and_recv` schedulers and NodeFlow SPMM schedulers now build a temporary (heterogeneous) bipartite graph (see changes in `python/dgl/runtime/spmv.py`)

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [x] Branch rebased to `master` after merging #749 (from which this PR is based)

## TODO

- [x] More tests
- [x] Check documentations

I intend to defer these to later PRs since they are less urgent, but I'm not sure (checked boxes are completed items):

- [x] `pop_n_repr` and `pop_e_repr`
- [ ] Default message/reduce/apply functions
- [ ] Better `__repr__`
- [ ] Subgraphs
- [x] Slice syntax (only on message passing)
- [ ] Conversion to/from networkx
- [ ] Conversion to/from DGLGraph.

### CODE REVIEW

**Comment by VoVAllen** [comment]
> We allow calling message passing methods (send, recv, pull, update_all, send_and_recv, etc.) as long as the subgraph contains only one edge type.

So we only support message passing on the graph with only one edge type?

**Comment by BarclayII** [comment]
> So we only support message passing on the graph with only one edge type?

Yes.  You can call message passing on different edge types one at a time though.

File: src/graph/bipartite.h [hunk 1]
```diff
@@ -171,6 +178,517 @@ class Bipartite : public BaseHeteroGraph {
   COOPtr coo_;
 };
 
+//////////////////////////////////////////////////////////
+//
+// COO graph implementation
+//
+//////////////////////////////////////////////////////////
```
Comment (jermainewang) [inline src/graph/bipartite.h]: Why move them in header file?

Comment (BarclayII) [reply src/graph/bipartite.h]: Because I need to include them in `src/kernel/binary_reduce.h`.

**Comment by BarclayII** [comment]
This is ready for review.

File: python/dgl/__init__.py [hunk 2]
```diff
@@ -17,6 +17,7 @@
 from .backend import load_backend
 from .batched_graph import *
 from .graph import DGLGraph
+from .heterograph import DGLHeteroGraph
```
[the remark this answers is not in the document]
Comment (BarclayII) [reply python/dgl/__init__.py]: Which means completely removing the concept of `DGLGraph` (i.e. treating it as a special case of `DGLHeteroGraph`)?

File: python/dgl/graph.py [hunk 3]
```diff
@@ -65,6 +73,10 @@ def is_readonly(self):
         """
         return self._graph.is_readonly()
 
+    def _number_of_edges(self):
+        """Return number of edges in the current view (only used for scheduler)"""
+        return self.number_of_edges()
```
[the remark this answers is not in the document]
Comment (BarclayII) [reply python/dgl/graph.py]: Yes.

File: python/dgl/graph.py [hunk 4]
```diff
@@ -939,6 +951,14 @@ def _get_msg_index(self):
...
         self._msg_index = index
 
+    @property
+    def _src_frame(self):
+        return self._node_frame
+
+    @property
+    def _dst_frame(self):
+        return self._node_frame
+
```
[the remark this answers is not in the document]
Comment (BarclayII) [reply python/dgl/graph.py]: I guess for bipartite graphs `node_frame` does not make sense at all.  According to the agreed API, node frames and edge frames should always be accessed from the heterograph itself (like `hg.ndata[ntype][feature]`.

File: python/dgl/heterograph.py [hunk 5]
```diff
@@ -882,96 +845,444 @@ def out_degree(self, v):
...
+            self._msg_indices[self._current_etype_idx] = utils.zero_index(
+                size=self._graph.number_of_edges(self._current_etype_idx))
+        return self._msg_indices[self._current_etype_idx]
+
+    def _set_msg_index(self, index):
+        self._msg_indices[self._current_etype_idx] = index
+
+    def __getitem__(self, key):
+        if key in self._etypes_invmap:
+            return self._create_view(None, self._etypes_invmap[key])
```
[the remark this answers is not in the document]
Comment (BarclayII) [reply python/dgl/heterograph.py]: Yes.

File: python/dgl/heterograph.py [hunk 6]
```diff
@@ -1,306 +1,307 @@
...
     def metagraph(self):
-        """Return the metagraph as networkx.MultiDiGraph."""
-        pass
-
-    def number_of_nodes(self):
-        """Return the number of nodes in the graph.
+        """Return the metagraph as networkx.MultiDiGraph.
+
+        The nodes are labeled with node type names.
+        The edges have their keys holding the edge type names.
```
[the remark this answers is not in the document]
Comment (BarclayII) [reply python/dgl/heterograph.py]: The node types would map to the node names in `nx.MultiDiGraph`.
The edge types would map to the _keys_ of edges in `nx.MultiDiGraph`.  One can access the keys via `nx_g.edges(keys=True)`.

File: python/dgl/heterograph.py [hunk 7]
```diff
@@ -1,306 +1,307 @@
...
+    @property
+    @utils.cached_member('_cache', '_current_dsttype_idx')
+    def _current_dsttype_idx(self):
+        """Checks the uniqueness of edge type in the view and get the index
+        of the destination type.
+
+        This allows reading/writing edge frame data and message passing routines.
+        """
+        _, dsttype_idx = self._endpoint_types(self._current_etype_idx)
+        return dsttype_idx
```
[the remark this answers is not in the document]
Comment (BarclayII) [reply python/dgl/heterograph.py]: Yes.

File: python/dgl/__init__.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jermainewang) [reply python/dgl/__init__.py]: Just a name suggestion.

**Comment by BarclayII** [comment]
Should we merge this for now?

**Comment by jermainewang** [comment]
@zheng-da any other comments?

**Comment by zheng-da** [comment]
LGTM

**Pull request merged**